DEVELOPMENT-GUIDE.md

```diff
@@ -0,0 +1,167 @@
+
+## How ACCP uses the JNI
+Just as in many other areas of development, there are a wide range of acceptable styles for JNI code.
+ACCP has adopted the following guidelines and patterns to make code more efficient as well as easier to write and review.
+
+### Avoid native calls to Java
+Crossing the JNI boundary in either direction is expensive. However, crossing it from Native to Java is far worse.
+For this reason, avoid calls that interact with `JNIEnv` (or, more properly, `raii_env` in ACCP).
+As a result:
+* Native code should never touch a Java object
+* Native code should never call a Java method
+* All parameters to native methods should be either primitives or primitive arrays.
+  (Technically, all arrays are Java objects and thus expensive to touch; however this is an unavoidable compromise).
+* Logic should use native data structures and only translate from and to Java results at the beginning and end of the top-most methods.
+* Native methods should all be `static`.
+  While this technically isn't necessary, if they aren't allowed to touch Java objects, then there is no need for them to be passed a reference to `this`.
+
+One result of this design decision is that ACCP's code can be (roughly) divided into four layers.
+These are not *currently* formally marked, but that may change in the future.
+
+Each layer from highest level (Java) down:
+1. Pure (normal) Java. This layer is called by external (non-ACCP) Java code.
+2. Java->Primitive Translation. This layer is written in Java and converts from Java objects to primitives and actually calls the `native` methods.
+3. JNI->Native Translation. This layer is written in C++ and converts from JNI objects to native or ACCP implemented structures (such as `java_buffer`).
+    This layer is also responsible for translating C++ exceptions thrown by the lowest layer to Java exceptions.
+4. Native. This layer is written in C++ and actually does the logic. It should rarely (if ever) directly interact with `JNIEnv`/`raii_env` and should rarely (if ever) throw a Java exception.
+
+### Error Handling
+Once a [Java Exception](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/design.html#java_exceptions) has been thrown, there are exceedingly few JNI operations a caller is allowed to perform.
+What's more, it is easy to not notice that there is a Java exception pending.
+So long as you do not interact with `JNIEnv`/`raii_env` directly, you shouldn't need to worry about Java exceptions as the ACCP objects/methods which use them already have appropriate checks which throw C++ exceptions (specifically `java_ex`) instead.
+If you need to throw an exception, it should almost always be a C++ exception and thrown using the `throw_java_ex` or `throw_openssl` methods. (The former is more efficient but the latter MUST be used when the exception is due to an error state reported by OpenSSL.)
```

Comment:
I've looked at an implementation of a JNI method and noticed that it catches only java_ex. This means C++ exceptions of other types will go to JVM and more likely will kill it, actual behaviour is undefined. So either all JNI methods must catch all C++ exceptions and convert them into Java exceptions if possible or only java_ex exceptions must be allowed. Anyway JNI methods must have the `catch(...) {}` block.

Reply:
Only `java_ex` are allowed.

I'll add explicit documentation around this.